component/xload/xload.go

```diff
@@ -0,0 +1,300 @@
+			if err != nil {
+				log.Err("Xload::Configure : config error creating directory of xload path [%s]", err.Error())
+				return fmt.Errorf("config error in %s [%s]", xl.Name(), err.Error())
+			}
+		}
+
+		if !common.IsDirectoryEmpty(xl.path) {
+			log.Err("Xload::Configure : config error %s directory is not empty", xl.path)
+			return fmt.Errorf("config error in %s [temp directory not empty]", xl.Name())
+		}
+	}
+
+	var mode xcommon.Mode
+	err = mode.Parse(conf.Mode)
+	if err != nil {
+		log.Err("Xload::Configure : Failed to parse mode %s [%s]", conf.Mode, err.Error())
+		return fmt.Errorf("invalid mode in xload : %s", conf.Mode)
+	}
+
+	if mode == xcommon.EMode.INVALID_MODE() {
+		log.Err("Xload::Configure : Invalid mode : %s", conf.Mode)
+		return fmt.Errorf("invalid mode in xload : %s", conf.Mode)
+	}
+
+	xl.mode = mode
+
+	return nil
+}
+
+// Start : Pipeline calls this method to start the component functionality
+func (xl *Xload) Start(ctx context.Context) error {
+	log.Trace("Xload::Start : Starting component %s", xl.Name())
+
+	xl.workerCount = xcommon.MAX_WORKER_COUNT
+	xl.blockPool = xcommon.NewBlockPool(xl.blockSize, xl.workerCount*3)
+	if xl.blockPool == nil {
+		log.Err("Xload::Start : Failed to create block pool")
+		return fmt.Errorf("failed to create block pool")
+	}
+
+	var err error
+
+	// Xload : start code goes here
+	switch xl.mode {
+	case xcommon.EMode.CHECKPOINT():
+		// Start checkpoint thread here
+		return fmt.Errorf("checkpoint is currently unsupported")
+	case xcommon.EMode.DOWNLOAD():
+		// Start downloader here
+		err = xl.startDownloader()
+		if err != nil {
+			log.Err("Xload::Start : Failed to start downloader [%s]", err.Error())
+			return err
+		}
+	case xcommon.EMode.UPLOAD():
+		// Start uploader here
+		return fmt.Errorf("uploader is currently unsupported")
+	case xcommon.EMode.SYNC():
+		//Start syncer here
+		return fmt.Errorf("sync is currently unsupported")
+	default:
+		log.Err("Xload::Start : Invalid mode : %s", xl.mode.String())
+		return fmt.Errorf("invalid mode in xload : %s", xl.mode.String())
+	}
+
+	return xl.startComponents()
+}
+
+// Stop : Stop the component functionality and kill all threads started
+func (xl *Xload) Stop() error {
+	log.Trace("Xload::Stop : Stopping component %s", xl.Name())
+
+	xl.comps[0].Stop()
+	xl.blockPool.Terminate()
+
+	// TODO:: xload : should we delete the files from local path
+	err := common.TempCacheCleanup(xl.path)
+	if err != nil {
+		log.Err("unable to clean xload local path [%s]", err.Error())
+	}
+
+	return nil
+}
+
+func (xl *Xload) startDownloader() error {
+	log.Trace("Xload::startDownloader : Starting downloader")
+
+	// Create remote lister pool to list remote files
+	rl, err := comp.NewRemoteLister(xl.path, xl.NextComponent())
+	if err != nil {
+		log.Err("Xload::startDownloader : Unable to create remote lister [%s]", err.Error())
+		return err
+	}
+
+	ds, err := comp.NewDownloadSplitter(xl.blockSize, xl.blockPool, xl.path, xl.NextComponent())
```

Comment:
If we provide blockpool as an object to the splitter, does it matter whether it knows the size of the block or not.

Reply:
Removed the block size parameter and reused the one set in block pool